pkg/controller/pxc/controller.go

```diff
@@ -343,115 +343,108 @@ func (r *ReconcilePerconaXtraDBCluster) Reconcile(request reconcile.Request) (re
+		if newHAProxyService.Spec.Type == corev1.ServiceTypeLoadBalancer || newHAProxyService.Spec.Type == corev1.ServiceTypeNodePort {
 			if len(o.Spec.HAProxy.ExternalTrafficPolicy) > 0 {
-				currentService.Spec.ExternalTrafficPolicy = o.Spec.HAProxy.ExternalTrafficPolicy
-			} else if currentService.Spec.ExternalTrafficPolicy != o.Spec.HAProxy.ExternalTrafficPolicy {
-				currentService.Spec.ExternalTrafficPolicy = corev1.ServiceExternalTrafficPolicyTypeCluster
+				newHAProxyService.Spec.ExternalTrafficPolicy = o.Spec.HAProxy.ExternalTrafficPolicy
+			} else if newHAProxyService.Spec.ExternalTrafficPolicy != o.Spec.HAProxy.ExternalTrafficPolicy {
+				newHAProxyService.Spec.ExternalTrafficPolicy = corev1.ServiceExternalTrafficPolicyTypeCluster
 			}
 		}
 
 		if o.CompareVersionWith("1.6.0") >= 0 {
-			currentService.Spec.Ports = append(ports,
+			newHAProxyService.Spec.Ports = append(ports,
 				corev1.ServicePort{
 					Port:       33062,
+					Protocol:   corev1.ProtocolTCP,
 					TargetPort: intstr.FromInt(33062),
 					Name:       "mysql-admin",
 				},
 			)
 		}
 
-		err = r.client.Update(context.TODO(), currentService)
+		err = updateServiceIfNeeded(r.client, oldHAProxyService, newHAProxyService)
 		if err != nil {
 			err = fmt.Errorf("HAProxy service upgrade error: %v", err)
 			return reconcile.Result{}, err
 		}
 
-		haProxyServiceReplicas := pxc.NewServiceHAProxyReplicas(o)
-		currentServiceReplicas := &corev1.Service{}
-		err = r.client.Get(context.TODO(), types.NamespacedName{Name: haProxyServiceReplicas.Name, Namespace: haProxyServiceReplicas.Namespace}, currentServiceReplicas)
+		currentHAProxyServiceReplicas := &corev1.Service{}
+		err = r.client.Get(context.TODO(), pxc.NewServiceHAProxyReplicasName(o), currentHAProxyServiceReplicas)
 		if err != nil {
 			return reconcile.Result{}, errors.Wrap(err, "failed to get HAProxyReplicas service")
 		}
 
+		newHAProxyServiceReplicas := currentHAProxyServiceReplicas.DeepCopy()
+
 		replicaPorts := []corev1.ServicePort{
 			{
 				Port:       3306,
+				Protocol:   corev1.ProtocolTCP,
 				TargetPort: intstr.FromInt(3307),
 				Name:       "mysql-replicas",
 			},
 		}
+
+		newHAProxyServiceReplicas.Spec.Ports = replicaPorts
+		newHAProxyServiceReplicas.Spec.Type = corev1.ServiceTypeClusterIP
 		if len(o.Spec.HAProxy.ReplicasServiceType) > 0 {
-			//Upgrading service only if something is changed
-			if currentServiceReplicas.Spec.Type != o.Spec.HAProxy.ReplicasServiceType {
-				currentServiceReplicas.Spec.Ports = replicaPorts
-				currentServiceReplicas.Spec.Type = o.Spec.HAProxy.ReplicasServiceType
-			}
-			//Checking default ServiceType
-		} else if currentServiceReplicas.Spec.Type != corev1.ServiceTypeClusterIP {
-			currentServiceReplicas.Spec.Ports = replicaPorts
-			currentServiceReplicas.Spec.Type = corev1.ServiceTypeClusterIP
+			newHAProxyServiceReplicas.Spec.Type = o.Spec.PXC.ReplicasServiceType
```

Comment:
Is it really should be PXC ReplicasServiceType, not HAProxy?

Reply:
56e4fe9c